upgrade_through_versions_test.py

```diff
@@ -970,60 +793,91 @@ def _multidc_schema_create(self):
-# where we will upgrade from the latest *tag* matching A, to the HEAD of branch B
-POINT_UPGRADES = []
-points = [v for v in UPGRADE_PATH if make_ver_str(v) >= '1.2']
-for i, _ in enumerate(points):
-    verslice = tuple(points[i:i + 2])
-    if len(verslice) == 2:  # exclude dangling version at end
-        POINT_UPGRADES.append(tuple(points[i:i + 2]))
-
-# create test classes for upgrading from latest tag on one branch, to head of the next branch (see comment above)
-for (from_ver, to_branch) in POINT_UPGRADES:
-    cls_name = ('TestUpgrade_from_' + make_ver_str(from_ver) + '_latest_tag_to_' + make_branch_str(to_branch) + '_HEAD').replace('-', '_').replace('.', '_')
-    from_ver_latest_tag = latest_tag_matching(from_ver)
-    debug('Creating test upgrade class: {} with start tag of: {} ({})'.format(cls_name, from_ver_latest_tag, sha_for_ref_name(from_ver_latest_tag)))
-    vars()[cls_name] = type(
-        cls_name,
-        (PointToPointUpgradeBase,),
-        {'test_versions': [from_ver_latest_tag, make_branch_str(to_branch)], '__test__': True})
-
-# create test classes for upgrading from HEAD of one branch to HEAD of next.
-for (from_branch, to_branch) in POINT_UPGRADES:
-    cls_name = ('TestUpgrade_from_' + make_branch_str(from_branch) + '_HEAD_to_' + make_branch_str(to_branch) + '_HEAD').replace('-', '_').replace('.', '_')
-    debug('Creating test upgrade class: {}'.format(cls_name))
-    vars()[cls_name] = type(
-        cls_name,
-        (PointToPointUpgradeBase,),
-        {'test_versions': [make_branch_str(from_branch), make_branch_str(to_branch)], '__test__': True})
-
-# create test classes for upgrading from HEAD of one branch, to latest tag of next branch
-for (from_branch, to_branch) in POINT_UPGRADES:
-    cls_name = ('TestUpgrade_from_' + make_branch_str(from_branch) + '_HEAD_to_' + make_branch_str(to_branch) + '_latest_tag').replace('-', '_').replace('.', '_')
-    to_ver_latest_tag = latest_tag_matching(to_branch)
-    # in some cases we might not find a tag (like when the to_branch is trunk)
-    # so these will be skipped.
-    if to_ver_latest_tag is None:
-        continue
-    debug('Creating test upgrade class: {} with end tag of: {} ({})'.format(cls_name, to_ver_latest_tag, sha_for_ref_name(to_ver_latest_tag)))
-
-    vars()[cls_name] = type(
+
+# used by function def immediately below
+created_class_scope = vars()
+
+
+def create_upgrade_class(version_list, protocol_version=DEFAULT_PROTOCOL_VERSION, bootstrap_test=False):
+    """
+    Dynamically creates a test subclass for testing the given versions.
+
+    'version_list' should be a list of versions ccm will recognize.
+    'bootstrap_test' is a boolean, if True bootstrap testing will be included.
+    """
+    if bootstrap_test:
+        parent_classes = (UpgradeTester, BootstrapMixin)
+    else:
+        parent_classes = (UpgradeTester,)
+
+    cls_name = 'TestUpgradeDynamicClass__' + "__".join(version_list).replace(':', '_').replace('-', '').replace('.', '_')
```

Comment:
Looked around out of curiosity: you may be able to avoid redefinition by adding a subclass hook to `UpgradeTester`:

https://docs.python.org/2/library/abc.html#abc.ABCMeta.__subclasshook__

Not sure how that interacts with manual subclassing with `type()`, but worth a try, just to make me feel better :P